### PR TITLE
Adds more info on XCMP and HRMP

### LINKED ISSUE #759: Document HRMP
Should go into XCMP page I think. There's some documentation here: https://github.com/paritytech/polkadot/pull/1503

### PR DESCRIPTION
**Pull request by lsaether**
closes #759

### CODE REVIEW

File: docs/glossary.md [hunk 1]
```diff
@@ -170,6 +170,14 @@ Defined by Jae Kwon of Cosmos as "a new chain that takes into account state from
 not to compete, but to provide broad access." A non-contentious blockchain that inherits the state
 of the underlying blockchain and creates a new branch of _the same blockchain_.
 
+## Horizontal Relay-routed Message Passing
+
+Horizontal Relay-routed Message Passing, also known as HRMP, is a precursor to the complete XCMP
+implementation that mimics the same interface and semantics of XCMP. It is similar to XCMP except
+for how it stores all messages in the Relay Chain storage, therefore making it more expensive and
+demaninding more resources than XCMP. It is planned to be retired once the implementation of XCMP is
```
Comment (krichard410) [inline docs/glossary.md]:
```suggestion
demanding more resources than XCMP. It is planned to be retired once the implementation of XCMP is
```

Comment (lsaether) [reply docs/glossary.md]: thanks!

File: docs/learn-crosschain.md [hunk 2]
```diff
@@ -26,7 +30,33 @@ messages.
...
+stores all messages in the Relay Chain storage. When XCMP has been implemented, HRMP is planned to
+be deprecated and phased out in favor of it.
+
+## Vertical Message Passing
+
+There are two kinds of vertical message passing, Upward Message Passing (UMP) and Downward Message
+Passing (DMP). UMP is used when a message originates on a parachain or a parathread to go from that
+parachain up to the Relay Chain. DMP is used to go the other way around, when a message originates
+from the Relay Chain and is destined for a parachain. Messages that are passed via DMP may originate
+from a parachain in which case, first UMP is used to communicate the message to the Relay Chain and
```
Comment (krichard410) [inline docs/learn-crosschain.md]:
```suggestion
from a parachain. In which case, first UMP is used to communicate the message to the Relay Chain and
```

File: docs/learn-crosschain.md [hunk 3]
```diff
@@ -26,7 +30,33 @@ messages.
 - Validators will check a proof that the new candidate for the next parachain block includes the
   processing of the expected ingress messages to that parachain.
 
-## Example
+XCMP queues must be initiated by first opening a channel between two parachains. The channel is
+identified by both the sender and recipeient parachains, meaning that it's a one-way channel. A pair
+of parachains can have at most two channels between them, one for sending messages to the other
+chain and another for receiving messages. Channel will require a deposit in DOT to be opened, which
```
Comment (krichard410) [inline docs/learn-crosschain.md]:
```suggestion
chain and another for receiving messages. The channel will require a deposit in DOT to be opened, which
```

File: docs/learn-crosschain.md [hunk 4]
```diff
@@ -26,7 +30,33 @@ messages.
...
+will get reciprocated when the channel is closed.
+
+## Horizontal Relay-routed Message Passing (HRMP)
+
+While XCMP is still being implemented, a stop-gap protocol known as HRMP exists in its place. HRMP
```
Comment (krichard410) [inline docs/learn-crosschain.md]: Perhaps here, could you elaborate on what a stop-gap protocol is? And maybe why this specific protocol is needed?

File: docs/learn-crosschain.md [hunk 5]
```diff
@@ -26,7 +30,33 @@ messages.
...
+has the same interface and functionality as XCMP but is much more demanding on resources since it
+stores all messages in the Relay Chain storage. When XCMP has been implemented, HRMP is planned to
+be deprecated and phased out in favor of it.
+
+## Vertical Message Passing
+
+There are two kinds of vertical message passing, Upward Message Passing (UMP) and Downward Message
```
Comment (krichard410) [inline docs/learn-crosschain.md]:
```suggestion
There are two kinds of Vertical Message Passing, Upward Message Passing (UMP) and Downward Message
```

File: docs/learn-crosschain.md [hunk 6]
```diff
@@ -60,3 +90,8 @@ will include this block for parachain B into the Relay Chain.
 
 - [XCMP Scheme](https://research.web3.foundation/en/latest/polkadot/XCMP.html) - Full technical
   description of cross-chain communication on the Web3 Foundation research wiki.
+- [Messaging Overview](https://w3f.github.io/parachain-implementers-guide/messaging.html) - An
+  overview of the messaging schemes from the Parachain Implementor's guide.
+- [XCMP Format](https://github.com/paritytech/xcm-format) - Description of the XCMP format.
+
+[xcmp-format]: https://github.com/paritytech/xcm-format
```
Comment (krichard410) [inline docs/learn-crosschain.md]:
```suggestion
[xcm-format]: https://github.com/paritytech/xcm-format
```
Not sure if you meant XCMP format or XCM format based on the link attached

File: docs/glossary.md [hunk 7]
```diff
@@ -170,6 +170,14 @@ Defined by Jae Kwon of Cosmos as "a new chain that takes into account state from
 not to compete, but to provide broad access." A non-contentious blockchain that inherits the state
 of the underlying blockchain and creates a new branch of _the same blockchain_.
 
+## Horizontal Relay-routed Message Passing
+
+Horizontal Relay-routed Message Passing, also known as HRMP, is a precursor to the complete XCMP
+implementation that mimics the same interface and semantics of XCMP. It is similar to XCMP except
```
Comment (Swader) [inline docs/glossary.md]:
```suggestion
implementation, that mimics the same interface and semantics of XCMP. It is similar to XCMP except
```

File: docs/glossary.md [hunk 8]
```diff
@@ -170,6 +170,14 @@ Defined by Jae Kwon of Cosmos as "a new chain that takes into account state from
...
+for how it stores all messages in the Relay Chain storage, therefore making it more expensive and
+demanding more resources than XCMP. It is planned to be retired once the implementation of XCMP is
```
Comment (Swader) [inline docs/glossary.md]: Consider rephrasing to: the plan is to retire it

File: docs/learn-crosschain.md (hunk 6 above)
Comment (lsaether) [reply docs/learn-crosschain.md]: Good catch!

**Comment by lsaether** [comment]
@krichard410 Thank you for the review. I've addressed your feedback.

File: docs/learn-crosschain.md [hunk 9]
```diff
@@ -26,7 +30,36 @@ messages.
...
+of parachains can have at most two channels between them, one for sending messages to the other
+chain and another for receiving messages. The channel will require a deposit in DOT to be opened,
+which will get reciprocated when the channel is closed.
+
+## Horizontal Relay-routed Message Passing (HRMP)
+
+While XCMP is still being implemented, a stop-gap protocol (see definition below) known as HRMP
+exists in its place. HRMP has the same interface and functionality as XCMP but is much more
+demanding on resources since it stores all messages in the Relay Chain storage. When XCMP has been
+implemented, HRMP is planned to be deprecated and phased out in favor of it.
```
Comment (krichard410) [inline docs/learn-crosschain.md:44]:
```suggestion
implemented, HRMP is planned to be deprecated and phased out in favour of it.
```

File: docs/learn-crosschain.md [hunk 10]
```diff
@@ -26,7 +30,36 @@ messages.
 - Validators will check a proof that the new candidate for the next parachain block includes the
   processing of the expected ingress messages to that parachain.
 
-## Example
+XCMP queues must be initiated by first opening a channel between two parachains. The channel is
+identified by both the sender and recipeient parachains, meaning that it's a one-way channel. A pair
+of parachains can have at most two channels between them, one for sending messages to the other
+chain and another for receiving messages. The channel will require a deposit in DOT to be opened,
+which will get reciprocated when the channel is closed.
```
Comment (Swader) [inline docs/learn-crosschain.md]: Is reciprocated the right word?

File: docs/learn-crosschain.md (hunk 9 above)
Comment (Swader) [reply docs/learn-crosschain.md:44]: We use American spelling so `favor` is OK

**Review by Swader — CHANGES_REQUESTED** [review]
Nits

**Pull request merged**